### PR TITLE
JAMES-2988 Compute read level given requested MessageProperties

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/MessagePropertiesTest.java [hunk 1]
```diff
@@ -87,4 +88,99 @@ void toOutputPropertiesShouldReturnHeadersMessagePropertyWhenIndividualHeadersRe
...
+        @Test
+        void combineShouldReturnHeaderWhenFullAndMetadata() {
+            assertThat(MessageProperties.ReadLevel.combine(
+                MessageProperties.ReadLevel.Metadata,
+                MessageProperties.ReadLevel.Full))
+                .isEqualTo(MessageProperties.ReadLevel.Full);
+        }
+
+        @Test
+        void combineShouldReturnHFullWhenFullAndHeader() {
```
Comment (rouazana) [inline server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/MessagePropertiesTest.java]: s/HFull/Full

**Comment by trantienduchn** [comment]
```
[62d0e5b4b31170b5508004fb1f3ef3df2911aebd] [ERROR]   GetMessagesMethodTest.processShouldReturnHeadersFieldWhenSpecificHeadersRequestedInPropertyList:390 » NoSuchElement
[62d0e5b4b31170b5508004fb1f3ef3df2911aebd] [ERROR]   GetMessagesMethodTest.processShouldReturnOnlyMandatoryPropertiesOnEmptyPropertyList:222 » NoSuchElement
[62d0e5b4b31170b5508004fb1f3ef3df2911aebd] [ERROR]   GetMessagesMethodTest.processShouldReturnPropertyFilterWhenFilteringHeadersRequested:416 » NoSuchElement
```

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/MessagePropertiesTest.java [hunk 2]
```diff
@@ -87,4 +88,99 @@ void toOutputPropertiesShouldReturnHeadersMessagePropertyWhenIndividualHeadersRe
...
+        @Test
+        void combineShouldReturnHeaderWhenHeaderAndMetadata() {
+            assertThat(MessageProperties.ReadLevel.combine(
+                MessageProperties.ReadLevel.Metadata,
+                MessageProperties.ReadLevel.Header))
+                .isEqualTo(MessageProperties.ReadLevel.Header);
+        }
+
+        @Test
+        void combineShouldReturnHeaderWhenFullAndMetadata() {
```
Comment (Arsnael) [inline server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/MessagePropertiesTest.java]: s/ShouldReturnHeader/ShouldReturnFull

**Comment by Arsnael** [comment]
```
[ERROR] spark.Spark - ignite failed
```
test this please

**Comment by chibenwa** [comment]
```
[02eb8e1752ba0eb008e06f89662b683a247e9cf7] [ERROR] Failed to execute goal on project apache-james-backends-jpa: Could not resolve dependencies for project org.apache.james:apache-james-backends-jpa:jar:3.5.0-SNAPSHOT: Failed to collect dependencies at com.google.guava:guava:jar:25.1-jre -> org.checkerframework:checker-qual:jar:2.0.0: Failed to read artifact descriptor for org.checkerframework:checker-qual:jar:2.0.0: Could not transfer artifact org.checkerframework:checker-qual:pom:2.0.0 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/org/checkerframework/checker-qual/2.0.0/checker-qual-2.0.0.pom.part (No such file or directory) -> [Help 1]
```

test this please

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/MessageProperties.java [hunk 3]
```diff
@@ -71,6 +71,28 @@ public MessageProperties toOutputProperties() {
...
+        // This defer from RFC-8621 behavior (not implemented here)
+        // If omitted, this defaults to: [ "partId", "blobId", "size", "name", "type", "charset", "disposition", "cid",
+        // "language", "location" ]
+        return readLevels.reduce(ReadLevel::combine)
+            .orElse(ReadLevel.Full);
+
+    }
+
+    private Stream<ReadLevel> headerPropertiesReadLevel() {
+        return headersProperties.map(collection ->
```
Comment (trantienduchn) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/MessageProperties.java]: the type of `headersProperties` makes me cringe

**Comment by chibenwa** [comment]
```
ConcurrentTestRunnerTest.runShouldPerformAllOperationsEvenOnOccasionalExceptions:230 » NotTerminated
```

test this please

**Comment by Arsnael** [comment]
CI restart...

test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/MessageProperties.java [hunk 4]
```diff
@@ -184,7 +211,25 @@ private static boolean outputProperty(MessageProperty p) {
             }
         }
     }
-    
+
+    public enum ReadLevel {
```
Comment (mbaechler) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/MessageProperties.java]: ReadProfile would be more appropriate

Comment (chibenwa) [reply server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/MessageProperties.java]: Thanks. See https://github.com/linagora/james-project/pull/2923

**Pull request closed without merging**